packages/useful-types/src/tests/types.test.ts

```diff
@@ -0,0 +1,233 @@
+import {spawnSync} from 'child_process';
+import path from 'path';
+
+import * as ts from 'typescript';
+
+import {Primitive} from '../types';
+
+function compileWithErrors(code: string) {
+  const filename = 'test.ts';
+  const script = `
+  import {ArrayElement} from 'types';
+
+  ${code}
+`;
+
+  const sourceFile = ts.createSourceFile(
+    filename,
+    script,
+    ts.ScriptTarget.Latest,
+  );
+
+  const defaultCompilerHost = ts.createCompilerHost({});
+
+  const customCompilerHost: ts.CompilerHost = {
+    getSourceFile: (name) => {
+      if (name === filename) {
+        return sourceFile;
+      } else {
+        return defaultCompilerHost.getSourceFile(name, ts.ScriptTarget.Latest);
+      }
+    },
+    writeFile: () => {},
+    getDefaultLibFileName: () => 'lib.d.ts',
+    useCaseSensitiveFileNames: () => false,
+    getCanonicalFileName: (filename) => filename,
+    getCurrentDirectory: () => '',
+    getNewLine: () => '\n',
+    getDirectories: () => [],
+    fileExists: () => true,
+    readFile: () => '',
+  };
+
+  const options = {
+    baseUrl: path.join(__dirname, '..'),
+    path: {
+      types: [path.join(__dirname, '..', 'types.ts')],
+    },
+  };
+
+  const program = ts.createProgram(['test.ts'], options, customCompilerHost);
+
+  const diagnostics = ts.getPreEmitDiagnostics(program);
+
+  const errors: string[] = [];
+
+  for (const diagnostic of diagnostics) {
+    if (diagnostic.file) {
+      const {line, character} = ts.getLineAndCharacterOfPosition(
+        diagnostic.file,
+        diagnostic.start!,
+      );
+      const message = ts.flattenDiagnosticMessageText(
+        diagnostic.messageText,
+        '\n',
+      );
+      errors.push(
+        `${diagnostic.file.fileName} (${line + 1},${
+          character + 1
+        }): ${message}`,
+      );
+    } else {
+      errors.push(
+        ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n'),
+      );
+    }
+  }
+
+  return errors;
+}
+
+function assertType<T>(input: T) {
+  expect(input).toStrictEqual(input);
+}
+
+enum NotMatch {
+  Value = 'no_match',
+}
+
+function isNoMatch(value: any): value is NotMatch.Value {
+  return value === NotMatch.Value;
+}
+
+function assertNotType<Actual, Expected>(
+  actual: Actual,
+  expected: Expected extends Actual ? NotMatch.Value : Expected,
+) {
+  if (typeof actual === typeof expected && !isNoMatch(expected)) {
+    throw new Error(
+      `cannot submit values of the same type. actual: ${actual}, expected: ${expected}`,
+    );
+  }
+  expect(actual).toStrictEqual(actual);
+}
+
+function extractCompilerErrors(code: string) {
```

Comment:
This approach just runs the snippet against ts-node with a custom tsconfig. The issue here is actually accessing the types defined in `./types` Will keep plucking away to get it to work.